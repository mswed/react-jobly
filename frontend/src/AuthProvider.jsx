import React, { useState, useEffect } from 'react';
import JoblyApi from './api';

// We create the context
const AuthContext = React.createContext();

// We set up a component to manage the context
const AuthProvider = ({ children }) => {
  // We initialize the token with the value from localStorage
  const [token, setToken] = useState(localStorage.getItem('jobly-token'));

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    JoblyApi.token = token;
    localStorage.setItem('jobly-token', token || '');
  }, [token]);

  const login = async (username, password) => {
    try {
      const token = await JoblyApi.login(username, password);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await JoblyApi.register(username, password, firstName, lastName, email);
      setToken(token);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  // Create the value object
  const value = { token, register };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
