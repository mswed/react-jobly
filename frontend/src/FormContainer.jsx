import { Container, Form, Button, Row, Col, Card } from 'react-bootstrap';

const FormContainer = ({ title, FormComponent }) => {
  return (
    <Container>
      <Row className="justify-content-center mt-5">
        <Col md={8} lg={6}>
          <h3 className="text-shadow">{title}</h3>
          <Card className="shadow-lg  p-3">
            <FormComponent />
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default FormContainer;
