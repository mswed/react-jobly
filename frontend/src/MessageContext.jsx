import React, { useState } from 'react';

// We create the context
const MessageContext = React.createContext();

// We set up a component to manage the context
const MessagesProvider = ({ children }) => {
  const [message, setMessage] = useState(null);

  const showMessage = (text, variant = 'success', duration = 5000) => {
    setMessage({ text, variant, duration });
    // Auto clear
    if (duration) {
      setTimeout(() => {
        clearMessage();
      }, duration);
    }
  };

  const clearMessage = () => {
    setMessage(null);
  };

  return <MessageContext.Provider value={{ message, showMessage, clearMessage }}>{children}</MessageContext.Provider>;
};

export { MessageContext, MessagesProvider };
