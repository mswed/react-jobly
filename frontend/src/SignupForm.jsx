import { Container, Form, Button, Row, Col, Card } from 'react-bootstrap';
import { useState, useContext } from 'react';
import { AuthContext } from './AuthProvider';
const SignupForm = () => {
  const INITIAL_STATE = {
    username: '',
    password: '',
    firstName: '',
    lastName: '',
    email: '',
  };

  // Get functions from AuthContext
  const { register } = useContext(AuthContext);

  // Manage the form inputs
  const [signupForm, setSignupForm] = useState(INITIAL_STATE);

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setSignupForm((originalData) => ({ ...originalData, [name]: value }));
  };

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    console.log('submitting form');
    await register(signupForm.username, signupForm.password, signupForm.firstName, signupForm.lastName, signupForm.email);
  };
  return (
    <Form onSubmit={handleSubmit}>
      <Form.Group className="mb-3" controlId="username">
        <Form.Label>Username</Form.Label>
        <Form.Control placeholder="Username" name="username" onChange={handleChange} value={signupForm.username} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="password">
        <Form.Label>Password</Form.Label>
        <Form.Control placeholder="Password" name="password" onChange={handleChange} value={signupForm.password} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="first-name">
        <Form.Label>First Name</Form.Label>
        <Form.Control placeholder="First Name" name="firstName" onChange={handleChange} value={signupForm.firstName} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="last-name">
        <Form.Label>Last Name</Form.Label>
        <Form.Control placeholder="Last Name" name="lastName" onChange={handleChange} value={signupForm.lastName} />
      </Form.Group>
      <Form.Group className="mb-3" controlId="email">
        <Form.Label>Email</Form.Label>
        <Form.Control placeholder="Email" name="email" onChange={handleChange} value={signupForm.email} />
      </Form.Group>
      <Button type="submit" className="w-100">
        Signup
      </Button>
    </Form>
  );
};

export default SignupForm;
