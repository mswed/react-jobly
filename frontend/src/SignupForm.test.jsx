import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import SignupForm from './SignupForm';
import { MemoryRouter } from 'react-router-dom';
import { MessagesProvider } from './MessageContext';
import { AuthProvider } from './AuthProvider';
import FormContainer from './FormContainer';

it('renders without crashing', async () => {
  render(
    <MemoryRouter>
      <MessagesProvider>
        <AuthProvider>
          <FormContainer title={'Login'} FormComponent={SignupForm} />
        </AuthProvider>
      </MessagesProvider>
    </MemoryRouter>
  );
});

it('matches snapshot', async () => {
  const { asFragment } = render(
    <MemoryRouter>
      <MessagesProvider>
        <AuthProvider>
          <FormContainer title={'Login'} FormComponent={SignupForm} />
        </AuthProvider>
      </MessagesProvider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
