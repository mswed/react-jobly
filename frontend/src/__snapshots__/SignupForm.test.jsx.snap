// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="justify-content-center mt-5 row"
    >
      <div
        class="col-lg-6 col-md-8"
      >
        <h3
          class="text-shadow"
        >
          Login
        </h3>
        <div
          class="shadow-lg  p-3 card"
        >
          <form
            class=""
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                name="username"
                placeholder="Username"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="first-name"
              >
                First Name
              </label>
              <input
                class="form-control"
                id="first-name"
                name="firstName"
                placeholder="First Name"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="last-name"
              >
                Last Name
              </label>
              <input
                class="form-control"
                id="last-name"
                name="lastName"
                placeholder="Last Name"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="email"
              >
                Email
              </label>
              <input
                class="form-control"
                id="email"
                name="email"
                placeholder="Email"
                value=""
              />
            </div>
            <button
              class="w-100 btn btn-primary"
              type="submit"
            >
              Signup
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
